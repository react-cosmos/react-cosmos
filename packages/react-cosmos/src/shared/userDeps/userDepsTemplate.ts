import path from 'path';
import slash from 'slash';
import { Json } from './shared';

type Args = {
  globalImports: string[];
  fixturePaths: string[];
  decoratorPaths: string[];
  rendererConfig: Json;
  rootDir: string;
  relativeToDir: string | null;
};
export function userDepsTemplate({
  globalImports,
  fixturePaths,
  decoratorPaths,
  rendererConfig,
  rootDir,
  relativeToDir,
}: Args) {
  const fixtures = createImportMap(fixturePaths, rootDir, relativeToDir);
  const fixtureKeys = Object.keys(fixtures);

  const decorators = createImportMap(decoratorPaths, rootDir, relativeToDir);
  const decoratorKeys = Object.keys(decorators);

  return `
// This file is automatically generated by Cosmos. Add it to .gitignore and
// only edit if you know what you're doing.

// Keeping global imports here is superior to making them bundle entry points
// because this way they become hot-reloadable.
${globalImports
  .map(p => `import '${resolveImportPath(p, relativeToDir)}';`)
  .join(`\n`)}

${fixtureKeys
  .map((k, i) => `import fixture${i} from '${fixtures[k]}';`)
  .join(`\n`)}

${decoratorKeys
  .map((k, i) => `import decorator${i} from '${decorators[k]}';`)
  .join(`\n`)}

export const rendererConfig = ${JSON.stringify(rendererConfig, null, 2)};

export const fixtures = {
${fixtureKeys.map((k, i) => `  '${k}': fixture${i}`).join(`,\n`)}
};

export const decorators = {
${decoratorKeys.map((k, i) => `  '${k}': decorator${i}`).join(`,\n`)}
};
`.trimLeft();
}

function createImportMap(
  paths: string[],
  rootDir: string,
  relativeToDir: string | null
): Record<string, string> {
  return paths.reduce(
    (acc, p) => ({
      ...acc,
      [slash(path.relative(rootDir, p))]: resolveImportPath(p, relativeToDir),
    }),
    {}
  );
}

function resolveImportPath(filePath: string, relativeToDir: string | null) {
  return relativeToDir
    ? `.${path.sep}${path.relative(relativeToDir, filePath)}`
    : filePath;
}
