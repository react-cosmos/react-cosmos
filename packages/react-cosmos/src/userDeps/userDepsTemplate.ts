import {
  userDepsImportMap,
  userDepsImportPath,
  UserDepsTemplateArgs,
} from './userDepsShared.js';

export function userDepsTemplate({
  globalImports,
  fixturePaths,
  decoratorPaths,
  rendererConfig,
  rootDir,
  relativeToDir,
}: UserDepsTemplateArgs) {
  const fixtures = userDepsImportMap(fixturePaths, rootDir, relativeToDir);
  const fixtureKeys = Object.keys(fixtures);

  const decorators = userDepsImportMap(decoratorPaths, rootDir, relativeToDir);
  const decoratorKeys = Object.keys(decorators);

  return `
// This file is automatically generated by Cosmos. Add it to .gitignore and
// only edit if you know what you're doing.

// Keeping global imports here is superior to making them bundle entry points
// because this way they become hot-reloadable.
${globalImports
  .map(p => `import '${userDepsImportPath(p, relativeToDir)}';`)
  .join(`\n`)}

${fixtureKeys
  .map((k, i) => `import fixture${i} from '${fixtures[k]}';`)
  .join(`\n`)}

${decoratorKeys
  .map((k, i) => `import decorator${i} from '${decorators[k]}';`)
  .join(`\n`)}

export const rendererConfig = ${JSON.stringify(rendererConfig, null, 2)};

const fixtures = {
${fixtureKeys
  .map((k, i) => ` '${k}': { module: { default: fixture${i} } }`)
  .join(`,\n`)}
};

const decorators = {
${decoratorKeys
  .map((k, i) => ` '${k}': { module: { default: decorator${i} } }`)
  .join(`,\n`)}
};

export const moduleWrappers = {
  lazy: false,
  fixtures,
  decorators,
};
`.trimStart();
}
