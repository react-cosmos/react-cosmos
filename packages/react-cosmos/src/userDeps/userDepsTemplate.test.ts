import { userDepsTemplate } from './userDepsTemplate';

const globalImports = [
  '/Users/ovidiu/cosmos/src/polyfills.ts',
  '/Users/ovidiu/cosmos/src/global.css',
];

const fixturePaths = [
  '/Users/ovidiu/cosmos/src/Counter/Counter.fixture.tsx',
  '/Users/ovidiu/cosmos/src/CounterButton/CounterButton.fixture.tsx',
];

const decoratorPaths = ['/Users/ovidiu/cosmos/src/cosmos.decorator.tsx'];

const rendererConfig = { port: 5002 };

it('should generate user deps module with absolute paths', () => {
  expect(
    userDepsTemplate({
      globalImports,
      fixturePaths,
      decoratorPaths,
      rendererConfig,
      rootDir: '/Users/ovidiu/cosmos/',
      relativeToDir: null,
    })
  ).toMatchSnapshot();
});

it('should generate user deps module with relative paths', () => {
  expect(
    userDepsTemplate({
      globalImports,
      fixturePaths,
      decoratorPaths,
      rendererConfig,
      rootDir: '/Users/ovidiu/cosmos',
      relativeToDir: '/Users/ovidiu/cosmos/src',
    })
  ).toMatchSnapshot();
});
