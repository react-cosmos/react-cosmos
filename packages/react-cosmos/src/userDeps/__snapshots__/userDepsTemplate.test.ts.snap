// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate TypeScript user deps module with absolute paths 1`] = `
"// This file is automatically generated by Cosmos. Add it to .gitignore and
// only edit if you know what you're doing.
import { RendererConfig, UserModuleWrappers } from 'react-cosmos-core';

import '/Users/ovidiu/cosmos/src/polyfills.ts';
import '/Users/ovidiu/cosmos/src/global.css';

import fixture0 from '/Users/ovidiu/cosmos/src/Counter/Counter.fixture.tsx';
import fixture1 from '/Users/ovidiu/cosmos/src/CounterButton/CounterButton.fixture.tsx';

import decorator0 from '/Users/ovidiu/cosmos/src/cosmos.decorator.tsx';

export const rendererConfig: RendererConfig = {
  "playgroundUrl": "http://localhost:5002"
};

const fixtures = {
 'src/Counter/Counter.fixture.tsx': { module: { default: fixture0 } },
 'src/CounterButton/CounterButton.fixture.tsx': { module: { default: fixture1 } }
};

const decorators = {
 'src/cosmos.decorator.tsx': { module: { default: decorator0 } }
};

export const moduleWrappers: UserModuleWrappers = {
  lazy: false,
  fixtures,
  decorators,
};
"
`;

exports[`should generate TypeScript user deps module with relative paths 1`] = `
"// This file is automatically generated by Cosmos. Add it to .gitignore and
// only edit if you know what you're doing.
import { RendererConfig, UserModuleWrappers } from 'react-cosmos-core';

import './polyfills.ts';
import './global.css';

import fixture0 from './Counter/Counter.fixture.tsx';
import fixture1 from './CounterButton/CounterButton.fixture.tsx';

import decorator0 from './cosmos.decorator.tsx';

export const rendererConfig: RendererConfig = {
  "playgroundUrl": "http://localhost:5002"
};

const fixtures = {
 'src/Counter/Counter.fixture.tsx': { module: { default: fixture0 } },
 'src/CounterButton/CounterButton.fixture.tsx': { module: { default: fixture1 } }
};

const decorators = {
 'src/cosmos.decorator.tsx': { module: { default: decorator0 } }
};

export const moduleWrappers: UserModuleWrappers = {
  lazy: false,
  fixtures,
  decorators,
};
"
`;

exports[`should generate user deps module with absolute paths 1`] = `
"// This file is automatically generated by Cosmos. Add it to .gitignore and
// only edit if you know what you're doing.

import '/Users/ovidiu/cosmos/src/polyfills.ts';
import '/Users/ovidiu/cosmos/src/global.css';

import fixture0 from '/Users/ovidiu/cosmos/src/Counter/Counter.fixture.tsx';
import fixture1 from '/Users/ovidiu/cosmos/src/CounterButton/CounterButton.fixture.tsx';

import decorator0 from '/Users/ovidiu/cosmos/src/cosmos.decorator.tsx';

export const rendererConfig = {
  "playgroundUrl": "http://localhost:5002"
};

const fixtures = {
 'src/Counter/Counter.fixture.tsx': { module: { default: fixture0 } },
 'src/CounterButton/CounterButton.fixture.tsx': { module: { default: fixture1 } }
};

const decorators = {
 'src/cosmos.decorator.tsx': { module: { default: decorator0 } }
};

export const moduleWrappers = {
  lazy: false,
  fixtures,
  decorators,
};
"
`;

exports[`should generate user deps module with relative paths 1`] = `
"// This file is automatically generated by Cosmos. Add it to .gitignore and
// only edit if you know what you're doing.

import './polyfills.ts';
import './global.css';

import fixture0 from './Counter/Counter.fixture.tsx';
import fixture1 from './CounterButton/CounterButton.fixture.tsx';

import decorator0 from './cosmos.decorator.tsx';

export const rendererConfig = {
  "playgroundUrl": "http://localhost:5002"
};

const fixtures = {
 'src/Counter/Counter.fixture.tsx': { module: { default: fixture0 } },
 'src/CounterButton/CounterButton.fixture.tsx': { module: { default: fixture1 } }
};

const decorators = {
 'src/cosmos.decorator.tsx': { module: { default: decorator0 } }
};

export const moduleWrappers = {
  lazy: false,
  fixtures,
  decorators,
};
"
`;
